src/vos/vos_internal.h

```diff
@@ -365,64 +365,79 @@ extern uint64_t vos_evt_feats;
 #define VOS_KEY_CMP_LEXICAL	(1ULL << 63)
 /** Indicates that a tree has aggregation optimization enabled */
 #define VOS_TF_AGG_OPT	(1ULL << 62)
-/** Indicates that the stored 32 bits from a timestamp are from an HLC */
+/** Indicates that the stored bits from a timestamp are from an HLC */
 #define VOS_TF_AGG_HLC	(1ULL << 61)
+/** Number of bits to use for timestamp (roughly 1/4 ms granularity) */
+#define VOS_AGG_NR_BITS	42
+/** HLC differentiation bits */
+#define VOS_AGG_NR_HLC_BITS	(64 - VOS_AGG_NR_BITS)
+/** Lower bits of HLC used for rounding */
+#define VOS_AGG_HLC_BITS	((1ULL << VOS_AGG_NR_HLC_BITS) - 1)
+/** Mask to check if epoch is HLC. */
+#define VOS_AGG_HLC_MASK	(VOS_AGG_HLC_BITS << (64 - VOS_AGG_NR_HLC_BITS))
+/** Mask of bits of epoch */
+#define VOS_AGG_EPOCH_MASK	(~VOS_AGG_HLC_MASK << (64 - VOS_AGG_NR_BITS))
 /** Start bit of stored timestamp */
 #define VOS_TF_AGG_BIT	60
-/** Right shift for 32-bits of epoch */
-#define VOS_TF_AGG_RSHIFT	(63 - VOS_TF_AGG_BIT)
-/** Left shift for 32-bits of epoch */
-#define VOS_TF_AGG_LSHIFT	(32 - VOS_TF_AGG_RSHIFT)
+/** Right shift for stored portion of epoch */
+#define VOS_AGG_RSHIFT	(63 - VOS_TF_AGG_BIT)
+/** Left shift for stored portion of epoch */
+#define VOS_AGG_LSHIFT	(VOS_AGG_NR_BITS - VOS_AGG_RSHIFT)
```

Comment:
Should this be "(64 - VOS_AGG_NR_BITS - VOS_AGG_RSHIFT)", that is, 19?

    [63, 61]: flags // RSHIFT?
    [60, 19]: stored portion of epoch
    [18,  0]: something else // LSHIFT?

Reply:
I need to add some tests for these corner cases.  I'll have a closer look.